cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java

```diff
@@ -234,54 +242,107 @@ public void dropAllForFlow(FlowId flowId) throws Exception {
 
   @Override
   public void upgrade() throws Exception {
-    // For each queue config table and queue data table in each namespace, perform an upgrade
-    for (final NamespaceMeta namespaceMeta : namespaceQueryAdmin.list()) {
-      impersonator.doAs(namespaceMeta.getNamespaceId(), new Callable<Void>() {
-        @Override
-        public Void call() throws Exception {
-          upgradeQueues(namespaceMeta);
-          return null;
+    int numThreads = 10; // TODO: Make this configurable
+    final ExecutorService executor =
+      Executors.newFixedThreadPool(numThreads,
+                                   new ThreadFactoryBuilder().setNameFormat("hbase-cmd-executor-%d").build());
+    try {
+      final Map<TableId, Future<?>> allFutures = new HashMap<>();
+      // For each queue config table and queue data table in each namespace, perform an upgrade
+      for (final NamespaceMeta namespaceMeta : namespaceQueryAdmin.list()) {
+        impersonator.doAs(namespaceMeta.getNamespaceId(), new Callable<Void>() {
+          @Override
+          public Void call() throws Exception {
+            Map<TableId, Future<?>> futures = upgradeQueues(namespaceMeta, executor);
+            allFutures.putAll(futures);
+            return null;
+          }
+        });
+      }
+
+      // Wait for the queue upgrades to complete
+      Map<TableId, Throwable> failed = waitForUpgrade(allFutures);
+      if (!failed.isEmpty()) {
+        for (Map.Entry<TableId, Throwable> entry : failed.entrySet()) {
+          LOG.error("Failed to upgrade queue table {}", entry.getKey(), entry.getValue());
         }
-      });
+        throw new Exception(String.format("Error upgrading queue tables. %s of %s failed",
+                                          failed.size(), allFutures.size()));
+      }
+    } finally {
+      try {
+        executor.shutdown();
+        executor.awaitTermination(5, TimeUnit.SECONDS);
```

Comment:
`call executor.shutdownNow with the threads as daemon thread so that the user can kill the upgrade at anything without blocking.` - I'll go with this approach